src/box/lua/config/applier/box_cfg.lua

```diff
@@ -428,7 +438,172 @@ local function apply(config)
     box.cfg(box_cfg)
 end
 
+local function missing_names_is_empty(missing_names, configdata)
+    return not missing_names[configdata._replicaset_name] and
+           table.equals(missing_names.peers, {})
+end
+
+local function cluster_on_replace(old, new)
+    -- Ignore insert of new replica. Every new replica will have
+    -- name set, as it may join to a replicaset only by bootstrap, which
+    -- means that config will pass instance_name to box.cfg. If user joins
+    -- replica without config module, we consider he understands, what he's
+    -- doing and don't automatically set instance_name.
+    if old == nil then
+        return
+    end
+
+    local is_replica_drop = old[3] == nil and new == nil
+    local is_name_insert = old[3] == nil and new and new[3] ~= nil
+    if is_name_insert or is_replica_drop then
+        box.on_commit(function()
+            worker:put('apply_names_ch', 'on_name_apply', {new[3]})
+        end)
+    end
+end
+
+local function schema_on_replace(old, new)
+    if old == nil and new and new[1] == 'replicaset_name' then
```

Comment:
I'd rather not do that. Even though the code seems to be almost ready for changing names, not only setting them, the problem is that if user doesn't start replica with force recovery option and master changes its name, then replica will fail to apply row and break replication. And I see no way to restore the cluster after that.

User should initiate name change from the replica with box.cfg and force_recovery enabled. In such case user will be stopped, if he forgets force_recovery, replication won't be broken.

At the end of a day, yes, we don't allow changing names, at least for now. In future, if somebody requests such feature it can be implemented: require UUID for replica, name of which is going to be changed, and force_recovery option. All name changes must be initiated from replicas itself, not from the master. Instances must be started without `instance/replicaset_name` passed to box.cfg in order to avoid `MISMATCH` error.